chunk/schema.go

```diff
@@ -0,0 +1,486 @@
+	PeriodicTableConfig
+	OriginalTableName string
+
+	// After midnight on this day, we start bucketing indexes by day instead of by
+	// hour.  Only the day matters, not the time within the day.
+	DailyBucketsFrom util.DayValue
+
+	// After this time, we will only query for base64-encoded label values.
+	Base64ValuesFrom util.DayValue
+
+	// After this time, we will read and write v4 schemas.
+	V4SchemaFrom util.DayValue
+}
+
+// RegisterFlags adds the flags required to config this to the given FlagSet
+func (cfg *SchemaConfig) RegisterFlags(f *flag.FlagSet) {
+	cfg.PeriodicTableConfig.RegisterFlags(f)
+
+	f.Var(&cfg.DailyBucketsFrom, "dynamodb.daily-buckets-from", "The date (in the format YYYY-MM-DD) of the first day for which DynamoDB index buckets should be day-sized vs. hour-sized.")
+	f.Var(&cfg.Base64ValuesFrom, "dynamodb.base64-buckets-from", "The date (in the format YYYY-MM-DD) after which we will stop querying to non-base64 encoded values.")
+	f.Var(&cfg.V4SchemaFrom, "dynamodb.v4-schema-from", "The date (in the format YYYY-MM-DD) after which we enable v4 schema.")
+}
+
+func (cfg *SchemaConfig) tableForBucket(bucketStart int64) string {
+	if !cfg.UsePeriodicTables || bucketStart < (cfg.PeriodicTableStartAt.Unix()) {
+		return cfg.OriginalTableName
+	}
+	// TODO remove reference to time package here
+	return cfg.TablePrefix + strconv.Itoa(int(bucketStart/int64(cfg.TablePeriod/time.Second)))
+}
+
+// componsiteSchema is a Schema which deletegates to various schemas depending
+// on when they were activated.
+type compositeSchema struct {
+	schemas []compositeSchemaEntry
+}
+
+type compositeSchemaEntry struct {
+	start model.Time
+	Schema
+}
+
+type byStart []compositeSchemaEntry
+
+func (a byStart) Len() int           { return len(a) }
+func (a byStart) Swap(i, j int)      { a[i], a[j] = a[j], a[i] }
+func (a byStart) Less(i, j int) bool { return a[i].start < a[j].start }
+
+func newCompositeSchema(cfg SchemaConfig) (Schema, error) {
+	schemas := []compositeSchemaEntry{
+		{0, v1Schema(cfg)},
+	}
+
+	if cfg.DailyBucketsFrom.IsSet() {
+		schemas = append(schemas, compositeSchemaEntry{cfg.DailyBucketsFrom.Time, v2Schema(cfg)})
+	}
+
+	if cfg.Base64ValuesFrom.IsSet() {
+		schemas = append(schemas, compositeSchemaEntry{cfg.Base64ValuesFrom.Time, v3Schema(cfg)})
+	}
+
+	if cfg.V4SchemaFrom.IsSet() {
+		schemas = append(schemas, compositeSchemaEntry{cfg.V4SchemaFrom.Time, v4Schema(cfg)})
+	}
+
+	if !sort.IsSorted(byStart(schemas)) {
+		return nil, fmt.Errorf("schemas not in time-sorted order")
+	}
+
+	return compositeSchema{schemas}, nil
+}
+
+func (c compositeSchema) forSchemas(from, through model.Time, callback func(from, through model.Time, schema Schema) ([]IndexEntry, error)) ([]IndexEntry, error) {
+	if len(c.schemas) == 0 {
+		return nil, nil
+	}
+
+	// first, find the schema with the highest start _before or at_ from
+	i := sort.Search(len(c.schemas), func(i int) bool {
+		return c.schemas[i].start > from
+	})
+	i--
```

Comment:
True; in reality I didn't think this would happen (as we stick a schema in at 0), but I guess someone could send a sample from before 1970 and crash us...